### DISCUSSION TITLE
v2.4

### DISCUSSION THREAD

**Discussion by eturpault** [GrandOrgue/OdfEdit#23] [Announcements]
30 May 2023
Changes compared to v2.3 :
- added a menu item checkbox to select the file format to use when saving an ODF (ISO_8859_1 or UTF-8 BOM)
- added a menu item checkbox to disable the automatic objects tree expand on object selection
- added a menu item checkbox to enable the wave based tremulants convertion from HW to GO
- added a menu item checkbox to enable the unused ranks convertion from HW to GO
- ctrl+s keys permits to save changes in the edited object and to save changed data in the ODF
- configuration data of the application (last ODF folder, options of the menu) are saved in a file OdfEdit.cfg
- added the possiblity to search a text inside the loaded HW ODF
- improvements made in the objects list/tree behavior on object selection or change
- objects having child(ren) cannot be deleted with the button Delete
- the parents/children of the selected object can be selected and edited in the central list (above the text editor)
- a double-click on a parent/child object makes it the selected object in the list/tree
- escape key permits to close the pop-up windows permitting to select parent/children objects
- HW2GO : removed the attribute AcceptRetuning=N in ranks of pipes
- HW2GO : pitch tuning attribute used for pipes which the sample has a different native frequency
- HW2GO : convert the conditional switches (switches which the state depends on the state of other switches)
- HW2GO : convert the synthetized and wave based tremulants
- HW2GO : the stop and coupler objects are numeroted according to the manual number to which they belong to
- several bugs fixing as at each release

Coming evolutions in next release : 
- manage drap-and-drop action on an object in the lists/tree to move it under another parent object
- add a manual compass extension feature
- suggestions of users are welcome

<hr /><em>This discussion was created from the release <a href='https://github.com/GrandOrgue/ODFEdit/releases/tag/v2.4'>v2.4</a>.</em>

**Comment by kerkovits** [comment]
I tried it, and got playable results. Good work!
In the following sample set:
https://www.sonusparadisi.cz/en/downloadable/download/sample/sample_id/15/
I encountered a problem: The clickable area turned out to be larger than the image for some GUI elements. It seems that ODFEdit is confused when ImageSet→ImageWidthPixels and ImageHeightPixels is missing. A possible solution would be to check the exact size of the bitmap and clamp ImageSet→ClickableAreaBottomRelativeYPosPixels and ClickableAreaRightRelativeXPosPixels to it. However, it would slow down conversion, as you need to load a raster processing module to Python, and check every bitmap for which the size is not specified in the ODF.

**Reply by eturpault** [reply]
With this sample set using the exact size of the bitmap would not help to define precisely the mouse clickable area, since it is larger than the button clickable area. For example with the stop "Flauta 8" of the manual I in the Single Jamb panel, here is the bitmap : 
![simple_flauta_8_off](https://github.com/GrandOrgue/ODFEdit/assets/51839171/45f968d3-87ff-4132-90c0-8c00a57700c6)

Having ImageSet→ImageWidthPixels and ImageHeightPixels not defined in the HW ODF is not a problem for OdfEdit, however there is a bug in the calculation of the clickable area width and height that I just realized thanks to this sample set.
I calculate : 
`MouseRectWidth = ClickableAreaRightRelativeXPosPixels `
instead of 
`MouseRectWidth = ClickableAreaRightRelativeXPosPixels - ClickableAreaLeftRelativeXPosPixels`
and the same for the height.
I just made the fix, so now the clickable area width and heigh are actually the one defined in the HW ODF, but it is not enough, the clickable area is still larger than the button area, especially on the right and bottom sides. This is because it is defined like that in the HW ODF which has : 
```
ClickableAreaLeftRelativeXPosPixels=70
ClickableAreaRightRelativeXPosPixels=476
ClickableAreaTopRelativeYPosPixels=35
ClickableAreaBottomRelativeYPosPixels=245
```
looking at the bitmap with an image editor, the attributes should be rather : 
```
ClickableAreaLeftRelativeXPosPixels=73
ClickableAreaRightRelativeXPosPixels=409
ClickableAreaTopRelativeYPosPixels=39
ClickableAreaBottomRelativeYPosPixels=224
```

If I set manually the values below in the GO ODF the clickable area in GO is good : 
```
MouseRectLeft=73
MouseRectTop=39
MouseRectWidth=336
MouseRectHeight=185
```

So except the fix that I did in OdfEdit, I don't see what I can do more, there are wrong coordinates in the HW ODF which have to be fixed manually in the GO ODF. What do you think ?

Besides, in OdfEdit v2.4 I already read the image size in the bitmap files if the attributes ImageWidthPixels and ImageHeightPixels are not defined in the HW ODF. This is due to the fact that some sample sets don't define the panel width and heigth, I suppose that Hauptwerk adapts automatically the panel size to the position and dimensions of the images which are inside the panel. So I need to know the images dimensions to set in the GO ODF the panel size when it is not defined in the HW ODF.

**Reply by kerkovits** [reply]
In my opinion, ODFEdit should check whether the clickable area fits into the image size. I see two options, which are not mutually exclusive:

1. During Check ODF validity, raise an error if MouseRect is too large. Now ODFEdit checks the resulting file with no error message.
2. On HW2GO, check the real image size, and if the lower or right border is not within the image, change its value to the maximum permitted value.

**Reply by eturpault** [reply]
Look at the images in the folder OrganInstallationPackages\001810\registry_simple_jamb of the sample set (from which is extracted the image that I inserted in my previous comment), you will see that in this case the clickable area has to be smaller than the image size, not equal to the image size. 
With the fix that I did yesterday for OdfEdit v2.5 the GO ODF will contain exactly what is defined in the HW ODF, which I expect is coherent with the images size. I could add the check to ensure that the MouseRect is not larger than the image size, but in the case of this sample set it will not be sufficient, the MouseRect has to be smaller than the image size in both width and height, and for this OdfEdit cannot analyze the content of each image to detect the real coordinates of the button inside the bitmap.

**Reply by kerkovits** [reply]
You misunderstood me. If ClickableAreaRightRelativeXPosPixels or ClickableAreaBottomRelativeYPosPixels is larger than image dimensions, decrease them to image dimensions, and calculate MouseRect with these modified dimensions (keeping ClickableAreaLeftRelativeXPosPixels and ClickableAreaTopRelativeYPosPixels intact). In this way, all errors are eliminated.

**Reply by eturpault** [reply]
Indeed we are speaking about different topics, sorry. ClickableAreaRightRelativeXPosPixels or ClickableAreaBottomRelativeYPosPixels larger than image dimensions should not occur if the HW ODF is well designed, and probably Hautpwerk would reject such an ODF as GrandOrgue does  for the same kind of issue. Since we convert to GO format HW ODF which are running well with Hauptwerk, I suppose that they are correctly designed and without errors. 
Anyway I can add in OdfEdit the test that you suggest in order to ensure to have MouseRect dimensions not exceeding image dimensions for the buttons.

**Comment by larspalo** [comment]
The 2.4 version was kind of successful in converting the St. Carlo, Brescia organ from Sonus Paradisi. Kind of, because the original version output couldn't be loaded in GO due to a mis-calculation of the number of (referenced) bass pipes to the short octave (24 pipes available but start reference from number 5 without decreasing the number of pipes accordingly). The modern version could load and play but lack the necessary pitch tuning lines to fill out the gaps in the last octave (short octave original with stretched out samples that should form extrapolated pipes not present in the original). There are a few other instances where pitch tuning lines are necessary too. Also, there seems to be no keyboard noises present in the conversion, though they do exist in the sample set (they do exist in the .xml file).

But overall, it's a nice improvement over previous versions that totally failed the conversion! Also, the way the stop noises are handled deserves some extra credit! I'll take note of it and utilize the same (or at least a similar) method in the future.

**Reply by eturpault** [reply]
Thanks for this feedback and the interresting use case of this St. Carlo Brescia organ which has unusual ranks compositions that I didn't took into account in OdfEdit, a new bone to gnaw for me ! This sample set has no demo version unfortulately but it is not expensive, I will buy it in order to add in OdfEdit the management of this special ranks/pipes configuration. I will come back to you if I have questions about how to manage it.
About the stop noises handling, I was inspired by how Piotr Graboswski made for his GO sample sets.
I confirm you that for the moment OdfEdit does not convert the keyboard noises, I will do it one day but I think it is a low priority evolution compared to other more interresting features or improvements to do.

**Reply by eturpault** [reply]
I fixed the issue with the attribute Rank999PipeCount in the stops which was to high with the original version.
I removed from Stop and Rank objects all the optional attributes which are set at a default value (as you do for GOODF), this is good to lighten the ODF. So Rank999PipeCount can be not present in the ODF.
Here are the generated ODF (original and standard) with these changes : 
[Brescia, St.Carlo, multi release.zip](https://github.com/GrandOrgue/ODFEdit/files/11644239/Brescia.St.Carlo.multi.release.zip)

About the necessary pitch tuning lines, there is no information in this HW ODF to indicate for example that +100 cents of pitch have to be applied to the sample pipe/Ottava/042-f#.wav which sounds at the same frequency as the sample pipe/Ottava/041-f.wav. So it can be only done manually in the GO ODF according to me, else if you see another solution.

I have a question : when a pipe has several attacks samples or several releases samples with same MaxKeyPressTime value (like in this sample set), how GrandOrgue selects which attack or release sample to play at each key press/release ? Are they all played together or only one is played and is selected randomly among all the defined one ?

**Reply by larspalo** [reply]
> I have a question : when a pipe has several attacks samples or several releases samples with same MaxKeyPressTime value (like in this sample set), how GrandOrgue selects which attack or release sample to play at each key press/release ? Are they all played together or only one is played and is selected randomly among all the defined one ?

They are (actually one of them at a time is) selected randomly if nothing else is specified.

> there is no information in this HW ODF to indicate for example that +100 cents of pitch have to be applied to the sample pipe/Ottava/042-f#.wav which sounds at the same frequency as the sample pipe/Ottava/041-f.wav.

I'm by no means an expert on the HW xml format, but I think there is information about it, but it's so "hidden" in the .xml file. One can however find entries like:

```
<ObjectList ObjectType="pipe_soundengine01">
	<pipe_soundengine01>
		<PipeID>98242</PipeID>
		<RankID>982</RankID>
		<GeneratePalletSwitch>N</GeneratePalletSwitch>
		<ControllingPalletSwitchID>982420</ControllingPalletSwitchID>
		<NormalMIDINoteNumber>42</NormalMIDINoteNumber>
		<Pitch_Tempered_BaseTuningSchemeCode>4</Pitch_Tempered_BaseTuningSchemeCode>
		<Pitch_Tempered_RankBasePitch64ftHarmonicNum></Pitch_Tempered_RankBasePitch64ftHarmonicNum>
		<Pitch_Tempered_BaseTuningDeviation>100</Pitch_Tempered_BaseTuningDeviation>

```

That seems to indicate the 100 cent shift. One would need to match both the rank and the pipe. One problem is that the extra attacks seem to have one digit more at the end of the pipe id number (98242 is kind of the base in this case and you have then a 0, 1, 2 etc as in 982420, 982421 etc).

Thanks for the odfs. Now it loads ok, but the short octave mapping is not quite right. I've not spent so much time examining how the mapping works in the HW .xml, but the 040-E should map as 036-C, 042-F# as 038-D and 044-G# as 040-E. The Norrfjärden sample set has one way of doing it and the GO help suggests a few different solutions.

Anyway, thanks for your efforts with OdfEdit! I really do think it could be a very valuable tool for GO users.

**Reply by eturpault** [reply]
I am glad you think OdfEdit can be useful, thanks !

Indeed it looks like the HW attribute Pitch_Tempered_BaseTuningDeviation attribute could be copied directly into the GO attribute Pipe999PitchTuning (currently I don't use it in OdfEdit), unfortunately in this sample set all the occurrences of this attribute are defined for blower / stop / key noises, it is never defined for real pipes, so I still don't see data in this ODF permitting to know which pipes should have to be retuned unfortunately.

About the mapping, do you mean the mapping between MIDI notes and sample wav files ? If yes, in the HW xml it is done thanks to the link between Pipe_SoundEngine01 objects (which define the MIDI note) and Sample objects (which define the wav files).
In OdfEdit, once a HW ODF has been loaded, on the right of the window there is a tab "Hauptwerk objects" which permits to navigate through the objects of the loaded ODF.
For example for this sample set, you can follow this path (double-click on each object to see its children) : _General -> Division000002 -> Stop000004 -> StopRank000003 -> Rank000040 -> Pipe_SoundEngine01004042 (MIDI=42) -> Pipe_SoundEngine01_Layer040420 -> Pipe_SoundEngine01_AttackSample040420 -> Sample040420 (file=pipe/Ottava/042-f#.wav).
So the sample 042-f#.wav is actually mapped to the MIDI note 42.

**Reply by larspalo** [reply]
As I said, I'm not at all an expert on the HW xml format. But to me it seems that there is some kind of additive identification going on that perhaps could help identifying what to look for. You have the rank number 982 (the `<RankID>982`) first then the midi note number 42 (`<PipeID>98242`)and then comes the attack/release numbering after that with 1, 2 etc if existing. The 982420 seems to be used for the `<ControllingPalletSwitchID>` which links a few different things together.

About the mapping, I was thinking about the short octave configuration of the original instrument. I have no idea of how that kind of mapping is done in HW, but I do know how it's supposed to work and I know about a few ways how it's implemented in GO. In a short octave instrument the lowest key seem to be 040-E but that's actually sounding the 036-C and so on. This is not really important for most of the sample sets, but a few sample sets of historical organs do utilize it. I wouldn't worry too much about it for now. Focus on the main things to improve and you can return to it later when you have time for it.

The St Carlo Brescia from Sonus Paradisi is one of the few sample sets from others that I've actually bought and really use myself even when I have access to many good organs in real life! In my opinion it's very reasonably priced, and definitely worth the investment!

**Reply by kerkovits** [reply]
I successfully reverse-engineered the HW format for manual noises.

There are two kinds of Keyboard objects. Some are built using `KeyboardKey` objects. Each has a distinct `SwitchID`. Look in `SwitchLinkage`, where you find this ID as `SourceSwitchID`. Action noise has `EngageLinkActionCode`=4, release noise has `DisengageLinkActionCode`=4. Now you have the correct `DestSwitchID`. Look for the same `ControllingPalletSwitchID` you found in table `Pipe_SoundEngine01`. That pipe will have a `RankID`. It is enough to seek this rank only for the leftmost key. Convert the two ranks (attack / release) to a GO `Rank999` as usual, and you can assign it to the corresponding `Manual999` by listing it among its stops.

Another place to look for keyboard noises is `StopRank`, which is used if `KeyboardKey` is empty. `ControllingSwitchID` is an invisible switch, which is always engaged, and `DivisionID` must be the same division the keyboard controls. `ActionEffectCode` is 2 for attack, 3 for release. In this way the `RankID` is found directly.

A hybrid solution is via `DivisionInput` objects. The `SwitchID` listed here equals to the `ControllingPalletSwitchID` of the pipe for attack, the release sample is found by searching `SwitchLinkage`→`DestSwitchID` with `SourceSwitchID`=`DivisionInput`→ `SwitchID` and `DisengageLinkActionCode`=2.

**Reply by kerkovits** [reply]
By the way, 'convert unused ranks' seems ineffective. If I look for the `RankID` I found using the method above in the HW objects panel, it does not have `_GO_uid` regardless of the setting 'convert unused ranks'.

**Reply by eturpault** [reply]
'convert unused ranks' converts HW ranks to GO ranks which are not linked to GO stop because they are precisely not used in the GO ODF (due to missing link with a StopRank in the HW ODF). The goal of this option is to permit to link manually these ranks to stops in the GO ODF once it has been build by OdfEdit if one wants to do it. This option has no effect on the ranks of noises, only on the ranks of pipes, but I could include the ranks of noises if needed. It is the reason why you see the field _GO_uid empty for these noise ranks.

Thank you for the analysis about how to recover the keyboard noises, it will help me when I will implement their conversion. There are indeed two ways to describe a keyboard in the HW ODF, it is applicable also to convert the graphical part of the keyboards.

**Reply by kerkovits** [reply]
It would be good to have optional unused noise ranks, as I can link them manually until the automatic conversion is done. In https://github.com/GrandOrgue/ODFEdit/issues/24 you can see that I am not the only one, who waits for this feature implemented. :-)

**Reply by kerkovits** [reply]
Caveat: Augustine's organs do not use ranks directly. Each key is mapped to a `StopRank` (refrencing its number by `MIDINoteNumOfFirstMappedDivisionInputNode`) and the corresponding pipe of the noise rank is irregularly selected using `MIDINoteNumIncrementFromDivisionToRank`.

As the `KeyboardKey`→`SwitchLinkage`→`Pipe_SoundEngine01` and `DivisionInput`{`NormalMIDINoteNumber`}(→`SwitchLinkage`)→`Pipe_SoundEngine01` and `StopRank`{`MIDINoteNumOfFirstMappedDivisionInputNode`;`MIDINoteNumIncrementFromDivisionToRank`}→`Pipe_SoundEngine01`{`NormalMIDINoteNumber`} methods all reference pipes one by one instead of full ranks, it seems that the best general approach is not to convert these noise ranks at all. However, if a keyboard noise is detected, new, `[Stop999]` objects are created for the manuals, filling it up with noise pipes directly (without ranks).

**Reply by eturpault** [reply]
To have right now the unused noise ranks converted to GO ODF when the option is enabled, simply comment the lines #6584 and 6586 in OdfEdit.py v2.4.

**Reply by kerkovits** [reply]
Thank you, it works.

**Reply by eturpault** [reply]
@kerkovits : I implemented the keys noise conversion from HW to GO, you can test it by recovering the file OdfEdit.py in the branch v2.7beta.

**Reply by kerkovits** [reply]
I tested it, and it works.

**Reply by eturpault** [reply]
Fine, thanks for the feedback !

**Comment by rnditdev** [comment]
I have a possible feature: Extend ranks, taking the last pipe definition, and create new ones in increments of 100 in pitch. Useful to bring 56 keys manuals to 61, the most common key number in MIDI controllers. Thanks.

**Reply by eturpault** [reply]
Thanks for the proposal. This new feature is planned for a next release as indicated at the end of the release note of the v2.4 : "add a manual compass extension feature".
What I intend to implement is to ask the user to select the Manual(s) to extend and until which MIDI note number to make this extension (under or above the current compass).

In order to preserve the original tuning of organs which have unequal temperaments, I think that it is better to make the extension by retuning pipes by one octave (so to use PitchTuning=1200) instead of by less than octave (PitchTuning=100 for example), has someone a opinion on this ?

**Comment by oleg68** [comment]
> I think that it is better to make the extension by retuning pipes by one octave (so to use PitchTuning=1200) instead of by less than octave (PitchTuning=100 for example), has someone a opinion on this ?

I agree totally.

**Comment by 25Tim** [comment]
ODFEdit.exe (V2.4) make Path Image=..\OrganInstallationPackages\.... and other, this don`t work in GrandOrgue. You can make the correct Path in ODFEdit (Image=OrganInstallationPackages\....)?

**Reply by eturpault** [reply]
Which error do you get in GrandOrgue to say that "this don't work" ? Which OS are you using ?
You are the first one to report such an issue. 
As indicated in the help of GO : "File paths are relative to the location of the ODF." The ODF being located in the folder OrganDefinitions of the Hauptwerk sample set, the relative path to access the folder OrganInstallationPackages which is at the same directory level is actually ..\OrganInstallationPackages, and it works well.

**Comment by 25Tim** [comment]


**Reply by eturpault** [reply]
Please share here in a file all the logs generated by OdfEdit when converting the HW ODF to GO ODF and by GO when loading your ODF. Since OdfEdit v2.0 the files path have been always generated in the same way and loaded without issue until your report.
Have you unzipped all the files of the sample set archive ?

**Comment by 25Tim** [comment]


**Reply by eturpault** [reply]
I don't see the pictures, but what you describe to solve your issue leads to think that your ODF is located in the folder where are the OrganDefinitions and OrganInstallationPackages folders, and not inside the OrganDefinitions folder as OdfEdit places it.
